### PR TITLE
Override deductive annotations

### PR DESCRIPTION
**Pull request by ignaspa**
## Motivation and Context

Graders should be able to switch between manually input grades and automatically calculated grades for marks when annotations with deductions for that mark could be applied.

## Your Changes

**Description**:

- Changing update mark route in results controller to take override
- Changing Delete button in flexible criterion mark element in results view.
- Changing data sent from controller to view when marks are updated
- Adding a revert_to_automatic_deductions route to results controller
- Change default value mark gets changed to in update_deduction when calculate_deduction returns 0. (so that max_mark is never automatically given)

**Type of change** (select all that apply):

 

- [x] New feature (non-breaking change which adds functionality)

## Testing

- Added tests for update_mark in results controller
- Added tests for revert_to_automoatic_deductions in results controller
- Modified tests to expect nil rather than max_mark when annotations changed in mark model spec
- UI testing of "revert to automatic deductions" and "delete mark" buttons

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have added tests for my changes, if applicable.
- [x] I have fixed any Hound bot comments (check after opening pull request).
- [x] I have verified that the TravisCI tests have passed (check after opening pull request).
- [x] I have reviewed the test coverage changes reported on Coveralls (check after opening pull request).

### CODE REVIEW

File: app/models/mark.rb [hunk 1]
```diff
@@ -34,13 +34,14 @@ def calculate_deduction
...
     deduction = calculate_deduction
+    return self.update!(mark: nil) if deduction == 0
     self.update!(mark: deduction > self.markable.max_mark ? 0.0 : self.markable.max_mark - deduction)
   end
 
   def ensure_mark_value
     return unless previous_changes.key?('override')
-    if previous_changes['override'].first == true && previous_changes['override'].second == false
-      update_deduction
+    if previous_changes['override'].second == false
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/models/mark.rb:43]: @david-yz-liu should I make a pull request to disable this cop?

File: spec/models/annotation_spec.rb [hunk 2]
```diff
@@ -38,7 +38,7 @@
 
     it 'correctly updates the mark when destroyed' do
       result.annotations.find_by(annotation_text: annotation_text).destroy
-      expect(mark.mark).to eq(3.0)
+      expect(mark.mark).to eq(nil)
```
Comment (david-yz-liu) [inline spec/models/annotation_spec.rb:41]: Good, but also add a second test when there are *multiple* deductive applied to the same result for the same criterion, and one of them is deleted. (In this case the updated mark should be based on the remaining deduction.)

File: app/controllers/results_controller.rb [hunk 3]
```diff
@@ -551,7 +560,11 @@ def update_mark
         num_marked = assignment.get_num_marked(nil)
       end
       render json: {
-        num_marked: num_marked
+        num_marked: num_marked,
+        new_mark_value: result_mark.reload.mark,
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/controllers/results_controller.rb:564]: What do you mean bu "save one update line in the `result.jsx` callbacks"? I still need to update the mark value correct?

[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/controllers/results_controller.rb]: Oh yes I misread that part, you're right that you'll still need to update the mark value. Good catch.

**Pull request merged**